mu4/extensions/internal/extensionunpacker.cpp

```diff
@@ -0,0 +1,172 @@
+
+    qint64 totalZipSize = 0;
+    for (const MQZipReader::FileInfo& fileInfo : zipFile.fileInfoList()) {
+        totalZipSize += fileInfo.size;
+    }
+
+    Ret freeSpace = hasFreeSpace(destination, totalZipSize);
+    if (!freeSpace) {
+        return freeSpace;
+    }
+
+    RetVal2<QString, QVersionNumber> meta = extensionMeta(&zipFile);
+    if (!meta.ret) {
+        return meta.ret;
+    }
+
+    QString extensionId = meta.val1;
+    QString extensionDirPath(destination + "/" + extensionId);
+    QVersionNumber version = meta.val2;
+
+    Ret actualVersion = checkActualVersion(destination, extensionId, version);
+    if (!actualVersion) {
+        bool previousVersionError = actualVersion.code() == static_cast<int>(Err::UnpackPreviousVersionExists);
+        if (previousVersionError) {
+            Ret remove = removePreviousVersion(extensionDirPath);
+            if (!remove) {
+                return remove;
+            }
+        } else {
+            return actualVersion;
+        }
+    }
+
+    Ret unzipExtension = unzip(&zipFile, extensionDirPath);
+    return unzipExtension;
+}
+
+Ret ExtensionUnpacker::isDirectoryWritable(const QString& directoryPath) const
+{
+    QFileInfo destinationDirInfo(directoryPath);
+    if (!destinationDirInfo.isWritable()) {
+        return make_ret(Err::UnpackDestinationReadOnly);
+    }
+
+    return make_ret(Err::NoError);
+}
+
+Ret ExtensionUnpacker::hasFreeSpace(const QString& directoryPath, qint64 neededSpace) const
+{
+    QStorageInfo destinationStorageInfo(directoryPath);
+    if (neededSpace > destinationStorageInfo.bytesAvailable()) {
+        return make_ret(Err::UnpackNoFreeSpace);
+    }
+
+    return make_ret(Err::NoError);
+}
+
+RetVal2<QString, QVersionNumber> ExtensionUnpacker::extensionMeta(const MQZipReader* zip) const
+{
+    QByteArray mdba = zip->fileData("metadata.json");
+
+    QJsonDocument loadDoc = QJsonDocument::fromJson(mdba);
+    QJsonObject mdObject = loadDoc.object();
+    QString extensionId = mdObject["id"].toString();
+    QVersionNumber version = QVersionNumber::fromString(mdObject["version"].toString());
+
+    mu::RetVal2<QString, QVersionNumber> result;
+    if (extensionId.isEmpty() || version.isNull()) {
+        result.ret = make_ret(Err::UnpackInvalidStructure);
+        return result;
+    }
+
+    result.ret = make_ret(Err::NoError);
+    result.val1 = extensionId;
+    result.val2 = version;
+
+    return result;
+}
+
+Ret ExtensionUnpacker::checkActualVersion(const QString& destination, const QString& extensionId,
+                                          const QVersionNumber& version) const
+{
+    QDir destinationDir(destination);
+    QStringList dirList = destinationDir.entryList(QStringList(extensionId), QDir::Dirs | QDir::NoDotAndDotDot);
+
+    if (!dirList.contains(extensionId)) {
+        return make_ret(Err::NoError);
+    }
+
+    QString actualExtensionMetaFilePath = destination + "/" + extensionId + "/" + "metadata.json";
```

Comment:
Better to place this code in configuration

Reply:
No, it is an independent class, it does not know about the other services.